Micro-optimisations for matmul 2.0: Electric boogaloo

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76828
* __->__ #75197
* #77735

This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by @ezyang, and it incurs on most of the complexity of the
code that this PR adds.

lezcano committed:
Micro-optimisations for matmul 2.0: Electric boogaloo

This PR implements the bulk of
https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in
https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by @ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

ezyang commented:
should we wait for https://github.com/pytorch/pytorch/pull/75195 before reviewing this?

Lezcano commented:
The changes are mostly independent, and this could be landed without the previous one, so I'd say it's fine to review this one now. 

I split everything into a stacj of PRs to isolate the CI errors. Note that I did not properly test what I pushed at the end, hence the red CI. It was mostly green before that change I believe.

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1596,13 +1596,13 @@ Tensor& vdot_out(const Tensor& self, const Tensor& other, Tensor& result) {
     if (dim_tensor1 == 3 && dim_tensor2 == 2 &&
-        tensor1.stride(-1) != 1 && tensor1.stride(0) == tensor1.size(1) * tensor1.size(2)) {
+        tensor1.strides().back() != 1 &&
+        tensor1.strides().front() == tensor1.sizes().cbegin()[1] * tensor1.sizes().cbegin()[2]) {
```
If you're going to optimize like this, it's probably even better to fetch tensor.strides() only once and don't rely on the compiler to CSE it.

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1646,8 +1645,17 @@ Tensor _matmul_impl(
     const Tensor& tensor1,
```
The comment above here might need to get updated?

ezyang replied:
in particular you use matrix-vector for both the 1/2 and 2/1 cases now, right?

ezyang replied:
(this wasn't done in this PR, I guess it was earlier in the stack)

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1658,101 +1666,120 @@ Tensor _matmul_impl(
+                              : MaybeOwned<Tensor>::borrowed(tensor1);
+    const auto t2 = transpose && dim_tensor1 == 2
+                      ? MaybeOwned<Tensor>::owned(tensor1.t())
+                      : MaybeOwned<Tensor>::borrowed(transpose ? tensor1 : tensor2);
```
This is displeasingly asymmetric. What's going on here?

Lezcano replied:
I knowww. This used to annoy me as well. But I also think it's the correct logic. Let me see if it can be rewritten in a more symmetric way.

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1658,101 +1666,120 @@ Tensor _matmul_impl(
+    const auto t2 = transpose && dim_tensor1 == 2
+                      ? MaybeOwned<Tensor>::owned(tensor1.t())
+                      : MaybeOwned<Tensor>::borrowed(transpose ? tensor1 : tensor2);
+    // Invariant: t1->dim() >= 3 && (t2->dim() == 1 || t2->dim() == 2)
```
And I guess, also you can directly matrix multiply t1 and t2? (Because that's what the transposing is about)

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1658,101 +1666,120 @@ Tensor _matmul_impl(
-  } else if (should_fold_into_mm(tensor1, tensor2)) {
+  } else if (should_fold_into_mm(tensor1, tensor2) ||
+             should_fold_into_mm(tensor2, tensor1)) {
     // optimization: use mm instead of bmm by folding tensor1's batch into
```
This is technically inaccurate now as we may also fold tensor2's batch

Lezcano replied:
Note that this was being done in an else that would call this function recursively with the transposed arguments. We merge the two to avoid the second function call and having to recompute things.

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1658,101 +1666,120 @@ Tensor _matmul_impl(
-      return has_out ? out.set_(res) : res;
+    const auto sizes_1 = t1->sizes();
+    auto output_shape = DimVector(sizes_1.begin(), sizes_1.end() - 1);
+    const auto batch_product = c10::multiply_integers(output_shape);
```
It's probably worth pointing out that batch product here has ALREADY done the folding (otherwise this feels like an off by one error)

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1658,101 +1666,120 @@ Tensor _matmul_impl(
+    if (!has_out) {
+      if (t2_is_matrix) {
+        // FIXME This path always does an unnecessary copy when transpose == true as the returned
+        // result from BLAS is already C-transposed
```
Is this true? If returned `output` has transposed memory layout, then output.transpose_(-2,-1) will untranspose it, giving contiguous layout, making the contiguous() call a no-op.

Lezcano replied:
The thing is that all these operations always return contiguous tensors, so the contiguous point always stands.

Something that could be done is to call `mm_out` with a matrix with the right strides as to not need the `contiguous()` call, but this would make this function non-composite.

It's not clear to me what's the right solution to this issue.

ezyang replied:
oh I see what you mean; you're saying that we would need to call blas directly so we can get the transposed layout and then just return it directly (rather than calling blas ~> make it contiguous ~> make it transposed again) (so it's two copies, then!)

Lezcano replied:
I believe I tested this back in the day, and the juggling of strides that `mm` performs is able to avoid any copy (you need to allocate the space, but that's it). But yep to the toher things.

Lezcano replied:
Note that we don't need to call blas directly, we can just call `mm_out`, but that makes this op non-composite.

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1658,101 +1666,120 @@ Tensor _matmul_impl(
+        // Swap last two elements of output_shape
+        std::iter_swap(output_shape.end() - 2, output_shape.end() - 1);
+        at::native::resize_output(out, output_shape);
+        out.transpose_(-2, -1);
```
This looks sketchy, we're not supposed to edit the metadata for a user given out tensor, we have to respect what the user asked us to write into (unless a resize occurred).

Lezcano replied:
Discussing this here https://github.com/pytorch/pytorch/pull/75197#discussion_r843896392 resolving this one.

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1658,101 +1666,120 @@ Tensor _matmul_impl(
+        out.copy_(reshaped_out.view_as(out));
+      }
+      if (transpose_out) {
+        out.transpose_(-2, -1);
```
You untranspose it later, but the damage is done: another thread reading the metadata for this tensor could temporarily see the "altered" metadata here; and there wouldn't be expected to be a race if the other thread was only supposed to be writing into the data (which is what `out=` is supposed to do)

ezyang replied:
Or if you want to avoid it when the input strides match exactly, a reference to a possibly allocated tensor

Lezcano replied:
I'll push the changes in a second.

I did the change and tested that it is indeed working as expected since the tests pass, and the tests fail if you put a `TORCH_CHECK(!transpose_out);` inside the `if (!reshaped_out.is_alias_of(out)) {` if, the condition fires.

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1658,101 +1666,120 @@ Tensor _matmul_impl(
     }
     if (dim_tensor2 > 1) {
       output_shape.push_back(p);
     }
```
These cases are always true now, no?

ezyang replied:
Actually maybe this is not true but then the comment at the top of this block is misleading

Lezcano replied:
I guess you refer to the comment
>     // We are multiplying b1 x n x m1 by b2 x m2 x p (where b1 and b2 can be lists);

That's fair, I'll edit that one.

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1658,101 +1666,120 @@ Tensor _matmul_impl(
-    std::array<int64_t, 3> tensor1_bmm_view = {expand_batch_product, n, m1};
-
-    std::array<int64_t, 3> tensor2_bmm_view = {expand_batch_product, m2, p};
+    const int64_t n = dim_tensor1 > 1 ? tensor1.sizes().cend()[-2] : 1LL;
```
I must admit, what's the point to having `tensor1.size(-2)` if you can't use it in code like here. Can we modify the definition of `size(...)` so that it inlines into the same code here? You basically want a non-bounds-checked version of sizes...

ezyang replied:
cc @swolchok

ezyang replied:
By the way I don't think cend() is saving you an arithmetic op because ArrayRef only stores pointer and length, so to calc cend you still end up doing pointer + length - 2

Lezcano replied:
I just use `cend` to be able to use negative indices really.

Fwiw, I believe the problem with `size()` is that it is a virtual function. I don't know how possible would it be to make it non-virtual, but it'd certainly make all this code more readable.

ezyang replied:
we should be able to devirtualize it. If you are interested in continuing to do microoptimizations like this let's talk about what this would entail.

Lezcano replied:
I wouldn't mind at all, but I think for now the priority is in working with @mruberry on primtorch, so this will have to wait. Now, it looks like the kind of task that @peterbell10 would like. Otherwise I can look into it in the future.

peterbell10 replied:
See #75416

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1658,101 +1666,120 @@ Tensor _matmul_impl(
+      if (t2_is_matrix) {
+        // FIXME This path always does an unnecessary copy when transpose == true as the returned
+        // result from BLAS is already C-transposed
+        const auto output = at::_unsafe_view(t1_folded.mm(*t2), output_shape);
```
I'd love to know why it's kosher to use unsafe_view here

Lezcano replied:
I used it because the previous code was using it as well. I believe that `_unsafe_view` shall be used when the input tensor is an xvalue, or that's what I got at least from the comments around that function.

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1664,101 +1666,119 @@ Tensor _matmul_impl(
-
-      Tensor res = res_T.reshape(shape).contiguous();
-      return has_out ? out.set_(res) : res;
+    const auto t1_folded = t1->reshape({folded_dim1, sizes_1.back()});
```
this is potentially a copy if t1 was transposed in line 1677, so we'd be better off calling bmm here.

Lezcano replied:
So, the perfect heuristic for this path would be "this reshape could be made into a view". That is equivalent to "t1.strides().back() == 1 && t1 is contiguous in memory", right?
I believe the second part of this can be checked by doing the following:
Let `st = t1.strides()`, `sz = t1.sizes()`. We need to check if `i = argmax(st)`, `st[i]*sz[i] == t1.numel()`. Is this right?
Incidentally, this is a generalisaiton of the current check of "should_fold"
https://github.com/pytorch/pytorch/blob/f6b2582a3de01f836bdfc398dd7014319a2fed38/aten/src/ATen/native/LinearAlgebra.cpp#L1610-L1617
Is this about right?

ngimel replied:
Yes I think so. Generally we should avoid "optimizations" that lead to copies, see e.g. https://github.com/pytorch/pytorch/blob/92d10decc448871cfbed258ad0ad670a9bd37047/aten/src/ATen/native/LinearAlgebra.cpp#L1704 that a user complains about #76702

ngimel commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1664,101 +1666,119 @@ Tensor _matmul_impl(
+        // FIXME This path always does an unnecessary copy when transpose == true as the returned
+        // result from BLAS is already C-transposed
+        const auto output = at::_unsafe_view(t1_folded.mm(*t2), output_shape);
+        return transpose ? output.transpose_(-2, -1).contiguous() : output;
```
again, as soon as you incur additional copy 3d-2d becomes not much of an optimization, bmm is likely better.

Lezcano replied:
We could still go through this path if we called `mm_out` with a matrix transposed strides, but that'd make `matmul` a non-composite.
Should we just dispatch to `bmm` in this case?

ngimel replied:
Yeah dispatching to bmm sounds good, and same should happen for `_out` cases, having different paths for `out` and non-out is super confusing.

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

Lezcano commented:
@ngimel fixing those cases to avoid a copies in the next PR in the stack.

ngimel commented:
ok, but can we remove different paths for `out` and non out versions in this one, it's very confusing.

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1626,34 +1627,35 @@ bool should_fold_into_mm(const Tensor& tensor1, const Tensor& tensor2) {
+- If both Tensors are 1-dimensional, (1d) the dot product (scalar) is returned.
+- If the arguments are 2D - 1D or 1D - 2D, the matrix-vector product is returned.
+- If both arguments are 2D, the matrix-matrix product is returned.
+- If one of the arguments is ND with N >= 3 and the other is 1D or 2D, and some
```
This is accurate description of semantics, we perform the op via mm or mv, but return tensor is still ND

ngimel commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1664,101 +1666,119 @@ Tensor _matmul_impl(
-
-    // fold the batch into the first dimension
-    // Why not tensor1.view(-1, size1[size1.size() -1])?
+  } else if (should_fold(tensor1, dim_tensor2) || should_fold(tensor2, dim_tensor1)) {
```
so I'd suggest we don't try to handle 2d-3d case now, given all the extra copies that it incurs it's better to avoid this, and it'll also make this PR simpler.

Lezcano replied:
All this is fixed in the next PR of the stack, so I suggest we leave it as it is, as this PR is not changing the inner dispatch of the operation, it just tries to do the same thing as the code before but faster.

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

ngimel reviewed (approved):
Can you please rebase to get CI signal?

Lezcano commented:
I'm starting to think that all this `test_jit` failures may not be unrelated, as I just rebased it on `viable/strict`... I'll look further into them on Monday

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul 2.0: Electric boogaloo"


This PR implements the bulk of https://github.com/pytorch/pytorch/pull/64387

Part of the optimisations were already merged in https://github.com/pytorch/pytorch/pull/72230

A number of these optimisations include:
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of these, it fixes a correctness issue of `matmul_out`, where the
out parameter was not resized correctly when passed to the backends.
This involves removing the use of `set_` from the calling code, as
requested by ezyang, and it incurs on most of the complexity of the
code that this PR adds.

[ghstack-poisoned]
